modules/300-prometheus/images/grafana/entrypoint/main.go

```diff
@@ -0,0 +1,179 @@
+		if err == nil {
+			err := cp.Copy(bundledPluginsPath, gfPathsPlugins)
+			if err != nil {
+				log.Fatalf("copy plugins: %v", err)
+			}
+		}
+	}
+
+	gfPathsConfig := os.Getenv("GF_PATHS_CONFIG")
+	if err := unix.Access(gfPathsConfig, unix.R_OK); err != nil {
+		log.Fatalf("GF_PATHS_CONFIG='%s' is not readable.\nYou may have issues with file permissions, more information here: http://docs.grafana.org/installation/docker/#migrate-to-v51-or-later", gfPathsConfig)
+	}
+
+	gfPathsData := os.Getenv("GF_PATHS_DATA")
+	if err := unix.Access(gfPathsData, unix.W_OK); err != nil {
+		log.Fatalf("GF_PATHS_DATA='%s' is not writable.\nYou may have issues with file permissions, more information here: http://docs.grafana.org/installation/docker/#migrate-to-v51-or-later", gfPathsData)
+	}
+
+	gfPathsHome := os.Getenv("GF_PATHS_HOME")
+	if err := unix.Access(gfPathsHome, unix.R_OK); err != nil {
+		log.Fatalf("GF_PATHS_HOME='%s' is not readable.\nYou may have issues with file permissions, more information here: http://docs.grafana.org/installation/docker/#migrate-to-v51-or-later", gfPathsHome)
+	}
+
+	err = convertEnv()
+	if err != nil {
+		log.Fatalf("convert env: %v", err)
+	}
+
+	os.Setenv("HOME", gfPathsHome)
+
+	err = installPlugins(gfInstallPlugins, gfPathsPlugins)
+	if err != nil {
+		log.Fatalf("convert env: %v", err)
+	}
+
+	grafanaArgs := []string{
+		"grafana-server",
+		fmt.Sprintf("--homepath=%s", gfPathsHome),
+		fmt.Sprintf("--config=%s", gfPathsConfig),
+		"--packaging=docker",
+		"cfg:default.log.mode=console",
+		fmt.Sprintf("cfg:default.paths.data=%s", gfPathsData),
+		fmt.Sprintf("cfg:default.paths.logs=%s", gfPathsLogs),
+		fmt.Sprintf("cfg:default.paths.plugins=%s", gfPathsPlugins),
+		fmt.Sprintf("cfg:default.paths.provisioning=%s", gfPathsProvisioning),
+	}
+
+	grafanaBin := "/usr/share/grafana/bin/grafana-server"
+
+	err = unix.Exec(grafanaBin, grafanaArgs, os.Environ())
+	if err != nil {
+		log.Fatalf("exec %s: %v", grafanaBin, err)
+	}
+}
+
+const (
+	paramPrefix = "GF_"
+	paramSuffix = "__FILE"
+)
+
+// convertEnv Convert all environment variables with names ending in __FILE into the content of
+// the file that they point at and use the name without the trailing __FILE.
+// This can be used to carry in Docker secrets.
+func convertEnv() error {
+	for _, param := range os.Environ() {
+		if !strings.HasPrefix(param, paramPrefix) {
+			continue
+		}
+		splitedParam := strings.Split(param, "=")
+		if !strings.HasSuffix(splitedParam[0], paramSuffix) {
+			continue
+		}
+		newParamName := strings.TrimRight(splitedParam[0], paramSuffix)
+		_, ok := os.LookupEnv(newParamName)
+		if ok {
+			return fmt.Errorf("error: both %s and %s are set (but are exclusive)", newParamName, splitedParam[0])
+		}
+
+		file, err := os.Open(splitedParam[1])
+		if err != nil {
+			log.Fatalf("open file: %v", err)
+		}
+		defer file.Close()
+
+		content, err := io.ReadAll(file)
+		if err != nil {
+			log.Fatalf("reading file: %v", err)
+		}
```

Comment:
```suggestion
		content, err := os.ReadFile(splitedParam[1])
		if err != nil {
			log.Fatalf("open file: %v", err)
		}
```

NIT: Deferring `Close` function calls in a loop can be a problem because the `Open` call happens immediately, and the `Close` call happens only when the functions exit after opening all the files.